lib/natalie/parser.rb

```diff
@@ -77,7 +77,9 @@ def ast
     end
 
     def encoding
-      result.source.source.encoding
+      encoding = result.magic_comments.find { it.key == 'encoding' }
+      return result.encoding if encoding.nil?
+      Encoding.find(encoding.value)
```

Comment:
This seems like an OK approach. I have to assume the change in https://github.com/ruby/prism/pull/3172 was for users of Prism as a library and not for language implementations. I'd like to see what TruffleRuby does with this. :popcorn: